app/models/email_message.rb

```diff
@@ -11,4 +11,23 @@ def self.find_for_reports(feedback_message_ids)
     select(:to, :subject, :content, :utm_campaign, :feedback_message_id)
       .where(feedback_message_id: feedback_message_ids)
   end
+
+  def self.fast_destroy_old_retained_email_deliveries(destroy_before_timestamp = 3.months.ago)
+    # We remove email delivery records periodically, except some we retain long term.
+    # We generally want to retain emails directly sent by human admins.
+    # The only email currently sent manually are those that are tied directly to a feedback message.
+    sql = <<-SQL.squish
+      DELETE FROM ahoy_messages
+      WHERE ahoy_messages.id IN (
+        SELECT ahoy_messages.id
+        FROM ahoy_messages
+        WHERE created_at < ? AND feedback_message_id = null
+        LIMIT 50000
```

Comment:
Question: why is there a LIMIT in this statement? Is there a concern that a large delete will lock the table?

Reply:
I copied exactly how it was done in https://github.com/forem/forem/blob/main/app/models/notification.rb#L138 and this seemed like a very similar technical challenge.

Part of my keeping this as draft for now is acknowledging that maybe it should be thought through more, but it kind of makes sense to do it this way, and if it needs to change, the functionality is encapsulated well.

One note on that front: I don't think the logic in the notification query is specifically tested (and maybe it's just too hard to test), but before this is merged I'd want for us to be sure we're good on what we test and how we test for this destructive functionality.

Reply:
@joshpuetz @benhalpern Oooh, I love this question! 🤩 There are actually multiple reasons to do it this way. Prepare yourself, this will almost certainly not be brief. :joy:

The biggest reason is that all work done inside a DB transaction is held in DB server memory until it's committed. DEV's DB has 256GB of RAM, so it can handle some pretty big transactions, but the first run of this operation will delete 49 million of the largest rows in the DB and will almost certainly invoke autovacuum when it's committed. When you consider that it's also not the only thing running on the DB at the time (DEV never really has _low_ traffic times and we run almost all of our background jobs during off-peak hours so even low-traffic times aren't east on the DB), we need to be considerate of other transactions running concurrently.

Doing all of this inside a single transaction will slow the rest of the DB operations to a crawl as it likely forces every other connection on the server into the swap while it accumulates this huge set of operations to commit to disk. So to avoid that, we want to perform this huge operation as a series of periodic commits (single statements are executed inside implicit transactions that are auto-committed when the statement completes — nearly every ACID-compliant DB does this) to release some of that memory throughout as we go along. Basically, it's [`find_in_batches`](https://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-find_in_batches), but for deletion and the batching is done on the DB side instead of the application.

To be clear, it's typically safe (for the DB, at least) to do huge `SELECT` queries, but not `INSERT`/`UPDATE`/`DELETE`, and I'm happy to go into detail on that if you like.

The second reason is that there's no way deleting that many rows will happen inside the [statement timeout](https://github.com/forem/forem/blob/c842d1a0c6718f319f167e9ec13d4d7d5d6b6d39/config/database.yml#L26) (the default is overridden to 10 seconds on DEV). This is timeout is enforced by Postgres rather than a `Timeout` call in Ruby.

If any of this is still muddy, I'm happy to clarify.